### PR TITLE
Feat/gamification of the app using localisation

### PR DESCRIPTION
**Pull request by kzepfl**
## Feature
When a user is in a radius of 30 meter from a park he would "discover the park" and receive 100 point.
And rating a park give you 20 point

### Point for entering new park
- `LocationService.kt`, new function `rewardParkDiscovery` for adding point when arriving in a new park
- `User.kt`/`UserRepository.kt`, add a List of visited park
- `UserRepository.kt`/`UserViewModel.kt`, add 2 function `getParksByUid` and `addNewPark`
- `MapUi.kt`, call the new function when position changes
- `Map.kt`, added log to check if the localisation stop

### Point when rating a park

- `ParkOverview.kt`, added `updateAndDisplayPoints` when rating

### Gamification

- `Progression.kt` added 2 new point : FIND_NEW_PARK(100) and  ADD_RATING(20)
- Added new argument for the Snackbard : navigationActions, scope and host in  `ParkOverview.kt` and `MapUi`
- `MainActivity.kt`, give the argument to  `ParkOverviewScreen` and `MapScreen`
- 
## Testing

- Update all test to new Park list argument in User MVVM
- Add new test in `UserViewModelTest.kt` : `addNewParkCallsRepositoryWithCorrectUids` and `getParksByUidCallsRepositoryWithCorrectUid`

### You can test the functionnality by seting your location inside a park
image

### CODE REVIEW

File: app/src/test/java/com/android/streetworkapp/model/user/UserViewModelTest.kt [hunk 1]
```diff
@@ -219,7 +352,15 @@ class UserViewModelTest {
   @Test
   fun getOrAddUserByUidUpdateUser() = runTest {
     val uid = "user123"
```
[the remark this answers is not in the document]
Comment (kzepfl) [reply app/src/test/java/com/android/streetworkapp/model/user/UserViewModelTest.kt:354]: I just changed the User to add the park argument, It was Paul's test.
I could make a global user, but user is different each test.

File: app/src/test/java/com/android/streetworkapp/model/user/UserRepositoryFirestoreTest.kt [hunk 2]
```diff
@@ -670,6 +745,49 @@ class UserRepositoryFirestoreTest {
...
+    assertNotNull(user)
+    assertTrue(user?.parks?.isEmpty() == true) // parks list should be empty
+  }
+
+  @Test
+  fun documentToUserWithExceptionWhenGettingParksSetsEmptyParksList() {
+    whenever(document.id).thenReturn("user123")
+    whenever(document.getString("username")).thenReturn("John Doe")
+    whenever(document.getString("email")).thenReturn("john@example.com")
+    whenever(document.getLong("score")).thenReturn(100L)
```
[the remark this answers is not in the document]
Comment (kzepfl) [reply app/src/test/java/com/android/streetworkapp/model/user/UserRepositoryFirestoreTest.kt:781]: I could but just like before these are Paul test and the whenever are sometime different.

```
  fun documentToUserWithMissingScoreSetsScoreToZero() {
    whenever(document.id).thenReturn("user123")
    whenever(document.getString("username")).thenReturn("John Doe")
    whenever(document.getString("email")).thenReturn("john@example.com")
    whenever(document.getLong("score")).thenReturn(null) // Missing score
    
      @Test
  fun documentToUserWithMissingFriendsSetsEmptyFriendsList() {
    whenever(document.id).thenReturn("user123")
    whenever(document.getString("username")).thenReturn("John Doe")
    whenever(document.getString("email")).thenReturn("john@example.com")
```

File: app/src/main/java/com/android/streetworkapp/utils/LocationService.kt [hunk 3]
```diff
@@ -3,14 +3,33 @@ package com.android.streetworkapp.utils
...
-class LocationService(private val context: Context) {
+class LocationService(
+    private val context: Context,
+    private val userViewModel: UserViewModel,
+    private val navigationActions: NavigationActions,
+    private val scope: CoroutineScope,
+    private val host: SnackbarHostState? = null
+) {
+  // Define the distance for when is in a park
+  private val insidePark = 30
```
[the remark this answers is not in the document]
Comment (kzepfl) [reply app/src/main/java/com/android/streetworkapp/utils/LocationService.kt:32]: fixed : 

...
companion object{
const val PARK_RADIUS = 30
}
...

**Review by misterM125 — APPROVED** [review]
Thank you for your changes!

**Pull request merged**